stlab/future.hpp

```diff
@@ -564,59 +572,59 @@ class future<T, detail::enable_if_copyable<T>> {
 
     future() = default;
 
-    bool valid() const { return static_cast<bool>(_p); }
+    bool valid() const { return static_cast<bool>(std::atomic_load(&_p)); }
```

Comment:
I thought that this is the way of doing a thread safe cancellation of futures. But I never before had the problem that a shared_ptr itself had to be handled in a thread safe manner. So it may be that my approach is completely wrong.